### PR TITLE
Live stream event tracking

### PR DESCRIPTION
**Pull request by justinswart**
This adds the tracking events for our live streams.  We track the following:

- ChangedLiveStreamOrientation
  - Device rotating to landscape/portrait while watching a live stream or replay.
- LiveStreamToggleSubscription
  - Subscribing/unsubscribing to live stream updates.
- ViewedLiveStreamCountdown
  - Viewed a live stream countdown.
- ViewedLiveStream
  - Viewed a live stream.
- WatchedLiveStream
  - Watched the live stream (pings every minute with the duration watched counted in minutes).
- ViewedLiveStreamReplay
  - Viewed a live stream replay.
- WatchedLiveStreamReplay
  - Watched a live replay (pings every minute with the duration watched counted in minutes).

The above are pretty straightforward and we have tests for tracking each minute now that we can travel through time in our testing.

We've also modified the `Viewed Project Page` event and added `live_stream_type`. This tracks whether or not a project had live streams when someone lands on a project page.  This is a little tricky due to how we receive the `liveStreams` array on the project. After a project page is viewed it performs a fetch for the full project data which includes `liveStreams` if any. If after 5 seconds there has been no `liveStreams` data returned it will track as not having live streams.

We also added a context to track live stream sharing via the `ShareViewModel`.

### CODE REVIEW

**Comment by mbrandonw** [comment]
btw this PR is dependent on https://github.com/kickstarter/ios-ksapi/pull/8

File: Library/Koala/Koala.swift [hunk 1]
```diff
@@ -804,6 +850,8 @@ public final class Koala {
     self.track(event: "Canceled Share Sheet",
                properties: props)
 
+    guard shareContext.isLiveStreamContext != true else { return }
```
Comment (mbrandonw) [inline Library/Koala/Koala.swift]: we're doing these `guard`s cause we don't do any live stream tracking for deprecated events... cant wait to get rid of deprecated events :/

File: Library/ViewModels/ProjectPamphletViewModel.swift [hunk 2]
```diff
@@ -86,11 +86,24 @@ ProjectPamphletViewModelOutputs {
...
-    Signal.combineLatest(project, refTag, cookieRefTag)
-      .takeWhen(self.viewDidAppearAnimated.signal)
+    // Try getting array of live streams from project, but if we can't after 5 seconds let's just emit `nil`
+    let projectLiveStreams: Signal<[Project.LiveStream]?, NoError> = project
+      .map { $0.liveStreams }
+      .skipNil()
+      .timeout(after: 5, raising: SomeError(), on: AppEnvironment.current.scheduler)
+      .materialize()
+      .map { $0.value }
+      .take(first: 1)
```
Comment (mbrandonw) [inline Library/ViewModels/ProjectPamphletViewModel.swift]: ^^^ this is the stuff justin was talking about in the description of this pr.

we want to track the `Viewed Project Page` event with a property that determines if there were any live streams available or not. unfortunately, we typically don't have that data right away... we have to wait until we get fresh data from the api. but, we also don't want to miss out on an event if that data never comes back (api error, network flakiness, etc...), so if after 5 seconds we get nothing we just default to `nil` and interpret that as not having any live stream data.

File: Library/Tests/ViewModels/ProjectPamphletViewModelTests.swift [hunk 3]
```diff
@@ -203,4 +207,130 @@ final class ProjectPamphletViewModelTests: TestCase {
     XCTAssertEqual(1, self.cookieStorage.cookies?.count,
                    "A single cookie has been set.")
   }
+
+  func testTracking_WaitingForLiveStreams_Timeout() {
```
Comment (mbrandonw) [inline Library/Tests/ViewModels/ProjectPamphletViewModelTests.swift]: ^ and here's where we are testing that timeout stuff...

File: Library/ViewModels/LiveStreamContainerViewModel.swift [hunk 4]
```diff
@@ -248,20 +231,72 @@ LiveStreamContainerViewModelInputs, LiveStreamContainerViewModelOutputs {
...
+          .trackChangedLiveStreamOrientation(project: project,
+                                             liveStream: liveStream,
+                                             toOrientation: orientation)
+    }
+
+    configData
+      .map { project, liveStream, _, refTag in (project, liveStream, refTag) }
+      .takePairWhen(numberOfMinutesWatched)
+      .map { tuple, minute in (tuple.0, tuple.1, tuple.2, minute) }
+      .take(during: Lifetime(self.token))
```
Comment (justinswart) [inline Library/ViewModels/LiveStreamContainerViewModel.swift]: We could achieve the same behaviour using a `takeUntil(until:)` and a `viewDidDisappear` signal, however as the issue here is that this signal never completes and we want its completion (and release) to be tied to the deinit of the viewModel, the `LifeTime` token felt more explicit.

**Comment by justinswart** [comment]
I think this is looking good! 👍

**Pull request merged**